File: Object-C/BlackFireIM/Example/ChatDetail/ChatRoom/BFChatRoomMemberListController.m
//
//  BFChatRoomMemberListController.m
//  BlackFireIM
//
//  Created by benny wang on 2021/10/29.
//

#import "BFChatRoomMemberListController.h"
#import "MSIMSDK-UIKit.h"
#import "BFGroupMemberCell.h"

@interface BFChatRoomMemberListController ()<UICollectionViewDelegate,UICollectionViewDataSource>

@property(nonatomic,strong) UICollectionView *myCollectionView;

@property(nonatomic,strong) NSMutableArray<MSGroupMemberItem *> *users;

@end

@implementation BFChatRoomMemberListController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.navView.navTitleL.text = @"Member List";
    [self.view addSubview:self.myCollectionView];
    [self loadData];
    [self addNotifications];
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter]removeObserver:self];
}

- (void)loadData
{
    [[MSIMManager sharedInstance]chatRoomMembers:self.roomInfo.room_id.integerValue successed:^(NSArray<MSGroupMemberItem *> * users) {
        [self.users removeAllObjects];
        [self.users addObjectsFromArray:users];
        [self.myCollectionView reloadData];
        
    } failed:^(NSInteger code, NSString *desc) {
        [MSHelper showToastFail:desc];
    }];
}

- (void)addNotifications
{
    WS(weakSelf)
    [[NSNotificationCenter defaultCenter]addObserverForName:MSUIKitNotification_ProfileUpdate object:nil queue:NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {
        [weakSelf profileUpdate: note];
    }];
    [[NSNotificationCenter defaultCenter]addObserverForName:MSUIKitNotification_ChatRoom_People_enter object:nil queue:NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {
        [weakSelf someOneEnter:note];
    }];
    [[NSNotificationCenter defaultCenter]addObserverForName:MSUIKitNotification_ChatRoom_People_leave object:nil queue:NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {
        [weakSelf someOneLeave:note];
    }];
}

- (UICollectionView *)myCollectionView
{
    if (!_myCollectionView) {
        UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc]init];
        layout.itemSize = CGSizeMake((Screen_Width-15*3)*0.5, (Screen_Width-15*3)*0.5*1.3);
        layout.sectionInset = UIEdgeInsetsMake(15 + NavBar_Height + StatusBar_Height, 15, 15 + Bottom_SafeHeight, 15);
        layout.minimumLineSpacing = 15;
        layout.minimumInteritemSpacing = 15;
        layout.scrollDirection = UICollectionViewScrollDirectionVertical;
        _myCollectionView = [[UICollectionView alloc]initWithFrame:CGRectMake(0, 0, Screen_Width, Screen_Height) collectionViewLayout:layout];
        _myCollectionView.delegate = self;
        _myCollectionView.dataSource = self;
        _myCollectionView.alwaysBounceVertical = YES;
        _myCollectionView.backgroundColor = [UIColor d_colorWithColorLight:TController_Background_Color dark:TController_Background_Color_Dark];
        _myCollectionView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
        [_myCollectionView registerClass:[BFGroupMemberCell class] forCellWithReuseIdentifier:@"memnberCell"];
    }
    return _myCollectionView;
}

- (NSMutableArray<MSGroupMemberItem *> *)users
{
    if (!_users) {
        _users = [NSMutableArray array];
    }
    return _users;
}

- (void)profileUpdate:(NSNotification *)note
{
//    NSArray *profiles = note.object;
    [self.myCollectionView reloadData];
}

- (void)someOneEnter:(NSNotification *)note
{
    MSGroupMemberItem *memeber = note.object;
    BOOL isExsit = NO;
    for (NSInteger i = 0; i < self.users.count; i++) {
        MSGroupMemberItem *item = self.users[i];
        if ([item.uid isEqualToString:memeber.uid]) {
            isExsit = YES;
            break;
        }
    }
    if (!isExsit) {
        [self.users addObject:memeber];
        [self.myCollectionView reloadData];
    }
}

- (void)someOneLeave:(NSNotification *)note
{
    NSString *uid = note.object;
    for (NSInteger i = 0; i < self.users.count; i++) {
        MSGroupMemberItem *item = self.users[i];
        if ([item.uid isEqualToString:uid]) {
            [self.users removeObject:item];
            [self.myCollectionView reloadData];
            break;
        }
    }
}

#pragma mark - UICollection view data source

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return self.users.count;
}

- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    BFGroupMemberCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"memnberCell" forIndexPath:indexPath];
    UILongPressGestureRecognizer *longPress = [[UILongPressGestureRecognizer alloc]initWithTarget:self action:@selector(userLongPress:)];
    [cell addGestureRecognizer:longPress];
    MSGroupMemberItem *item = self.users[indexPath.row];
    cell.info = item;
    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
 
}

- (void)userLongPress:(UILongPressGestureRecognizer *)ges
{
    switch (ges.state) {
        case UIGestureRecognizerStateBegan:
        {
            BFGroupMemberCell *cell = (BFGroupMemberCell *)ges.view;
            MSGroupMemberItem *item = cell.info;
            [self showMoreAction:item];
        }
            break;
        default:
            break;
    }
}

- (void)showMoreAction:(MSGroupMemberItem *)item
{
    if ([item.uid isEqualToString:[MSIMTools sharedInstance].user_id]) return;
    if (self.roomInfo.action_assign == NO && self.roomInfo.action_mute == NO) return;
    WS(weakSelf)
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    if (self.roomInfo.action_assign) {
        NSString *showTitle1 = item.role == 0 ? @"设置Ta为临时管理员" : @"取消Ta的管理员身份";
        [alert addAction:[UIAlertAction actionWithTitle:showTitle1 style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [weakSelf changeUserRole: item];
        }]];
    }
    if (self.roomInfo.action_mute) {
        NSString *showTitle2 = item.is_mute == NO ? @"禁言" : @"取消禁言";
        [alert addAction:[UIAlertAction actionWithTitle:showTitle2 style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [weakSelf changeUserMute: item];
        }]];
    }
    [alert addAction:[UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)changeUserRole:(MSGroupMemberItem *)item
{
    //当uid是正值时 是任命， 当为负值时是 取消任命
    WS(weakSelf)
    [[MSIMManager sharedInstance] editChatroomManagerAccess:self.roomInfo.room_id uids:@[item.role == 0 ? @(item.uid.integerValue) : @(-item.uid.integerValue)] duration:1 reason:@"good job!" successed:^{
        
        [MSHelper showToastSucc:@"Success"];
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [weakSelf.myCollectionView reloadData];
        });
        
    } failed:^(NSInteger code, NSString *desc) {
        [MSHelper showToastFail:desc];
    }];
}

- (void)changeUserMute:(MSGroupMemberItem *)item
{
    //当uid是正值是禁言，当uid为负值时是取消禁言
    WS(weakSelf)
    [[MSIMManager sharedInstance] muteMembers:self.roomInfo.room_id uids:@[item.is_mute ? @(-item.uid.integerValue) : @(item.uid.integerValue)] duration:1 reason:@"Don`t like a good guy" successed:^{
        
        [MSHelper showToastSucc:@"Success"];
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [weakSelf.myCollectionView reloadData];
        });
        
    } failed:^(NSInteger code, NSString *desc) {
        [MSHelper showToastFail:desc];
    }];
}

@end

File: Swift/MSIM-Swift/ChatDetail/ChatRoom/BFChatRoomMemberListController.swift
//
//  BFChatRoomMemberListController.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/11/2.
//

import UIKit
import MSIMSDK
import SwiftUI


public class BFChatRoomMemberListController: BFBaseViewController {

    public var roomInfo: MSGroupInfo!
    
    lazy var myCollectionView: UICollectionView = {
        let layout = UICollectionViewFlowLayout()
        layout.itemSize = CGSize(width: (UIScreen.width - 15 * 3) * 0.5, height: (UIScreen.width - 15 * 3) * 0.5 * 1.3)
        layout.sectionInset = UIEdgeInsets(top: 15, left: 15, bottom: 15, right: 15)
        layout.minimumLineSpacing = 15
        layout.minimumInteritemSpacing = 15
        layout.scrollDirection = .vertical
        let myCollectionView = UICollectionView(frame: CGRect(x: 0, y: 0, width: UIScreen.width, height: UIScreen.height), collectionViewLayout: layout)
        myCollectionView.dataSource = self
        myCollectionView.delegate = self
        myCollectionView.alwaysBounceVertical = true
        myCollectionView.backgroundColor = UIColor.d_color(light: MSMcros.TController_Background_Color, dark: MSMcros.TController_Background_Color_Dark)
        myCollectionView.register(BFGroupMemberCell.self, forCellWithReuseIdentifier: "userCell")
        return myCollectionView
    }()
    
    var dataArray: [MSGroupMemberItem] = []
    
    public override func viewDidLoad() {
        super.viewDidLoad()

        navigationItem.title = "Member List"
        view.addSubview(myCollectionView)
        
        loadData()
        addNotifications()
    }
    
    private func loadData() {
        if let roomID = self.roomInfo?.room_id,let room_id = Int(roomID) {
            MSIMManager.sharedInstance().chatRoomMembers(room_id) { users in
                
                self.dataArray.removeAll()
                self.dataArray += users
                self.myCollectionView.reloadData()
                
            } failed: { _, desc in
                MSHelper.showToastFailWithText(text: desc ?? "")
            }
        }
    }
    
    private func addNotifications() {
        NotificationCenter.default.addObserver(forName: NSNotification.Name.init(rawValue: MSUIKitNotification_ProfileUpdate), object: nil, queue: OperationQueue.main) {[weak self] note in
            self?.profileUpdate(note: note)
        }
    }
    
    @objc func profileUpdate(note: Notification) {
        
        self.myCollectionView.reloadData()
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}

extension BFChatRoomMemberListController: UICollectionViewDelegate,UICollectionViewDataSource {
    
    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.dataArray.count
    }

    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "userCell", for: indexPath) as! BFGroupMemberCell
        let longPress = UILongPressGestureRecognizer(target: self, action: #selector(userLongPress))
        cell.addGestureRecognizer(longPress)
        cell.info = self.dataArray[indexPath.row]
        return cell
    }
    
    @objc private func userLongPress(ges: UILongPressGestureRecognizer) {
        
        switch ges.state {
        case .began:
            if let cell = ges.view as? BFGroupMemberCell,let item = cell.info {
                showMoreAction(item: item)
            }
        default:
            break
        }
    }
    
    private func showMoreAction(item: MSGroupMemberItem) {
        
        let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        let showTitle1: String = item.role == 0 ? "设置Ta为临时管理员" : "取消Ta的管理员身份"
        alert.addAction(UIAlertAction(title: showTitle1, style: .default, handler: {[weak self] _ in
            self?.changeUserRole(item: item)
        }))
        let showTitle2: String = item.is_mute == false ? "禁言" : "取消禁言"
        alert.addAction(UIAlertAction(title: showTitle2, style: .default, handler: {[weak self] _ in
            self?.changeUserMute(item: item)
        }))
        alert.addAction(UIAlertAction(title: "取消", style: .cancel, handler: nil))
        self.present(alert, animated: true, completion: nil)
    }
    
    private func changeUserRole(item: MSGroupMemberItem) {
        
        if self.roomInfo.action_assign == false {
            MSHelper.showToastWithText(text: "exceed your authority")
            return
        }
        //当uid是正值时 是任命， 当为负值时是 取消任命
        let uid: Int = Int(item.uid)!
        MSIMManager.sharedInstance().editChatroomManagerAccess(self.roomInfo.room_id, uids: [item.role == 0 ? NSNumber(value: uid) : NSNumber(value: -uid)], duration: 1, reason: "good job!") {[weak self] in
            
            MSHelper.showToastSuccWithText(text: "Success")
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                self?.myCollectionView.reloadData()
            }
        } failed: { _, desc in
            MSHelper.showToastFailWithText(text: desc ?? "")
        }
    }
    
    private func changeUserMute(item: MSGroupMemberItem) {
        
        if self.roomInfo.action_mute == false {
            MSHelper.showToastWithText(text: "exceed your authority")
            return
        }
        //当uid是正值是禁言，当uid为负值时是取消禁言
        let uid: Int = Int(item.uid)!
        MSIMManager.sharedInstance().muteMembers(self.roomInfo.room_id, uids: [item.is_mute ? NSNumber(value: -uid) : NSNumber(value: uid)], duration: 1, reason: "Don`t like a good guy") {[weak self] in
            
            MSHelper.showToastSuccWithText(text: "Success")
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                self?.myCollectionView.reloadData()
            }
        } failed: { _, desc in
            MSHelper.showToastFailWithText(text: desc ?? "")
        }
    }
}
